usim/_primitives/task.py

```diff
@@ -153,17 +192,25 @@ def status(self) -> TaskState:
             return TaskState.CREATED
         return TaskState.RUNNING
 
-    def __close__(self, reason=TaskExit('activity closed')):
+    def __close__(self, reason=TaskClosed('activity closed')):
         """
         Close the underlying coroutine
 
         This is similar to calling :py:meth:`Coroutine.close`,
         but ensures that waiting activities are properly notified.
         """
+        # we have not FINISHED running yet, and can still change the result
         if self._result is None:
-            self.__runner__.close()
             self._result = None, reason
-            self._done.__set_done__()
+            # We have not STARTED running yet, which means we will in the same time
+            # Our __runner__ will cleanly exit as soon as it sees we have finalized
+            # everything already.
+            if self.__runner__.cr_frame.f_lasti == -1:
+                self._done.__set_done__()
+            # We are RUNNING and __runner__ is prepared to catch GeneratorExit
+            # Close the __runner__ to have it clean up and finalize everything.
```

Comment:
Wouldn't it be better to have the comments within the branches of the if statement? I was confused first why you could say in the outer scope that you did not start yet.